### ISSUE TITLE
[Learning] Cover Elxir Concepts in Crash Course

### ISSUE THREAD

**Issue by rtshkmr** [ve1ld/vyasa#5]
## Motivation 

To cover the example-driven dive into elixir. 
-- refer to wiki for learning artefacts: https://github.com/ve1ld/vyasa/wiki/%F0%9F%90%A6%E2%80%8D%F0%9F%94%A5-Elixir-&-Phoenix-Crash-Course-Notes-%F0%9F%90%A6%E2%80%8D%F0%9F%94%A5 

#### Source: 
* [Pragmatic Studio ](https://pragmaticstudio.com/signin)

**Comment by rtshkmr** [comment]
## Basic Course Takeaways & QQs 

1. QQ: saw that the video demo didn't need recompiling when a template file (.eex) was updated. Resulted in the question where I wondered if the templates are interpreted at runtime (since no recompile was necessary).
   * cursory look says that no. I'm going to move on from this question first.

**Comment by rtshkmr** [comment]
## Review of basic concepts with @KosmonautX 

1. the analogy of process spawning model in Elixir to organisms, cell death and all that. Understanding the overall aggregate behaviours by mapping it to the body itself.
2. Task mechanisms ==> good for deferred computations, where we don't need any blocking and we don't care about the response. Can do tasks async and spin it off to childprocs so that the msg queue doesn't get blocked. GenServer can end up blocking things (e.g. when the msg queue bloats up). 
3. rmb use of pin operator: 
  * rmb that overriding a var is actually creating a new variable. `pid = pid + 1`. This is still immutable actually. So the pin operator's purpose is to prevent the "mutation" and just to get a reference to value of the var. 
   * Good for db queries where there's a ref to a val as opposed to a var (as opposed to not using the ^ operator where it would have matched the variable instead of the value) 
   
   
 TODO: 
 1. understand how observer works. WX is the GUI (?) it's a CPP thing. Likely an Erlang implementation. 
   * Elixir used to have little support for GUIs, now it's improving

**Comment by rtshkmr** [comment]
__consider doing these for ex. :__  
1. Live Components 
2. 16: streams 
3. Js Commands and Js Hooks ==> interfacing with javascript
4. Live Navigation 
5. [19] Real time updates

**Comment by rtshkmr** [comment]
## Liveview Pro Course Notes  


WIP

**Comment by rtshkmr** [comment]
<details> 
<summary>Diagrams for LiveView Lifecycle</summary> 

![](https://online.pragmaticstudio.com/assets/courses/liveview-2ed-pro/lifecycle-1.jpg)

![](https://online.pragmaticstudio.com/assets/courses/liveview-2ed-pro/lifecycle-2.jpg)

![](https://online.pragmaticstudio.com/assets/courses/liveview-2ed-pro/lifecycle-3.jpg)

</details>

**Comment by rtshkmr** [comment]
## Ecto Required Knowledge  

Notes :
1. db migration files can be directly modified to effect the required db related changes
2. 

### Resources
1. [quick primer](https://blog.logrocket.com/getting-started-ecto-phoenix/) on ecto

**Comment by rtshkmr** [comment]
15. Streams: shows a way to handle large data volumes without relying on limited in-memory stores 
   * The goal is to do crud actions to large collections of state without needing to refetch the items from the database after the initial render AND ALSO the LiveView doesn't carry around large collections of items in its state. 
   * sockets will have collections and state changes would be streaming new entries to the named-socket in order to update that state held within a [stream](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#stream/4)
   * Some guidelines about using stream:
     * the browser tracks stream items via a `dom-id`, which is what the `streams` map (a parallel to the assigns map for in-mem store). This computed dom-id (which is the first value in the return tuple from the streams map) needs to be used to id the dom element that renders the stream item. 
     * Also, stream items need to be a parent dom container that wraps them. The container will have a unique id (can be anything) and have `phx-update="stream"` as it's binding. The binding helps show where new DOM elements (corresponding to stream items) should be appended / prepended 
     * for in-place updates to state (e.g. toggling state): 
       * allow the event to be emitted as usual and add a `handle_event` as usual. note that we can add value related bindings for events via `phx-value-*` which can allow us to nest values (e.g. identifier num [id]) that can be referenced to via the params in the event handler. this is how things are added to the event payload.
       * update the db: use the context module's functions from within the event handler (in the lifeview component). This shouldn't fail, and should update the database. 
       * handle the re-rendering for the updated stream-item. To do this, we want to change the stream in-place via the `stream_insert()` fn. Since that stream item already exists in the dom with that unique dom-id, then that dom element gets updated.[1]
   * performance benefit: after rendering, the stream item isn't kept in memory anymore [ "automatically kicked out of the stream"] (since it's already in the dom, with unique ids and they are contained in the server) 
     * this behaviour is similar to `temporary_assigns()`, but the key difference is whether the data is dynamic or static: 
       * Use temporary_assigns for static data that only ever needs to be rendered. Once it's rendered, it can safely be kicked out of the LiveView's state to conserve server-side memory. 
       * Use streams for collections of items that you don't want to hold in memory, but you need to update the collection after it's been initially rendered for example to append, prepend, update, or delete items.

<details><summary>[1] Here's an example event handler that handles state update of a stream item in-place</summary>
<img width="812" alt="image" src="https://github.com/ve1ld/vyasa/assets/38996397/bbfab056-9820-4704-833f-78ebb9928b34">
</details> 

16.  Live Components: a way to reuse the markup and its event handlers when attempting to externalise a piece of live view code.
  * without it, we'd need to duplicate both the markup (the heex template part) and the event handlers everywhere we reuse the LiveView components. 
  * rules: 
    * Live View components must be defined in their own module
    * needs to use the `:live_component` behaviour 
    * since the purpose is to render a heex component (with event handlers), the live view component also has a `render()` function. 
    * gotchas: 
      * live components must have a single static html tag at their root (e.g. a `<div> </div>` tag that encapsulates it)
      * any events emitted by a live component get send to the parent LiveView. Hence, there's a need to target the events emitted to go to the children (the live component)  -- `phx-target={@myself}` which is a binding to the `<.form/>` from within the live component whereby the `@myself` gets resolved to the current component instance which in this case is the child (live component).
  * calling that live component: `<.live_component module={<LiveViewComponentName>} id={<anythinghere e.g. :new>} />`
    * each live component on a page needs to be uniquely identifiable -- need to pass it an id attribute.
    * has to be unique amongst that particular type of live component.  different types of live components can have colliding IDs and they won't conflict. 
  * A live component can keep its own state as well, via the same callback functions (`mount()`, `render()`, handlers...)
    * but event-emitting needs to target the child live component explicitly via a phx-target (if not it defaults to looking into the parent)
    * but the child live component won't have access to the streams map since the streams map will be within the parent component. Therefore, the child needs to notify the parent component that the streams map state needs to be updated.
      * Any live-component rendered by a LiveView runs in the same process as its parent's LiveView --> the child live component can notify the parent live-view by sending a message to self() (which returns the PID of the parent LiveView). 
       So this requires the parent (LiveView component) to have a handle_info() that handles the message passed by the child (live component), and the logic is the same state update logic. 
  
17.

**Comment by rtshkmr** [comment]
17. PubSub! 
 * it's at the application-level. every application will have a PubSub service (which is a GenServer). Procs within that application can sub to topics and publish to topics. For subbed topics, handle_info hand
<img width="1794" alt="image" src="https://github.com/ve1ld/vyasa/assets/38996397/de74b677-ffa4-48a5-bf42-c5ae6136ff5a">

**Comment by rtshkmr** [comment]
18. Authenticating LiveViews 
 * auth code is kickstarted by Phoenix out of the box and will work. The generator is like so: `mix phx.gen.auth Accounts User users` where the `Accounts` context module uses the `User` schema module and the `users` db table.
 * random reminders: 
   * LiveViews can't set cookies in the browser because LiveView events are pushed over a websocket connection and for security reasons, cookies can't be set over a websocket. 
     * that's why session controllers are bare Phoenix controllers and not live views themselves. 
   * Some nomenclature: 

     * `pipeline`: a named sequence of plugs that run to-to-bottom.
     * `plug`: a function that takes a Phx connection (a struct containing data for the current HTTP request) and returns an updated connection -- it's a transformer of the connection struct -- central to how Phoenix handles HTTP requests. 
     *  `live`: a scope sub-definition whereby the appropriate pipeline is invoked. And only after the connection struct gets passed from the start to the end of the pipeline, will the struct be passed to the LiveView component. Note that adding a keyword list for the scope's `pipe_through` would mean that the connection struct traverses the pipelines in order of that keyword list. plugs can also be passed in and it would be equivalent to a pipeline. 
   *  the router is where the flow of routes get defined:
     * requests get plugged into live views via plugs (e.g. `:browser` and `:require_authenticated_user`). 
* so for auth, the flow goes like this:  <img width="1794" alt="image" src="https://github.com/ve1ld/vyasa/assets/38996397/4389d6cf-07c2-4310-96a4-718851a28ecd">
* gotcha: , if after running the phx.gen.auth generator you don't see the :fetch_current_user plug in router.ex then you'll need to manually add it after the :put_secure_browser_headers plug:
  ```elixir
  pipeline :browser do
    ...
    plug :put_secure_browser_headers
    plug :fetch_current_user
  end
  ``` 
* key takeaway: any auth checks must be done in 2 places: 
  * 1: when a user browses, directly via LiveView, via HTTP
  * 2: in the liveview's `mount()` lifecycle to account for live re-directs (via the `on_mount()` hook from within the LiveView module)

**Comment by rtshkmr** [comment]
closing because issallgoooddd
